jgrapht-core/src/test/java/org/jgrapht/alg/tour/TwoApproxMetricTSPTest.java

```diff
@@ -0,0 +1,183 @@
+                new TwoApproxMetricTSP<Object, DefaultEdge>().getTour(g);
+            assertHamiltonian(g, tour);
+
+            double mstWeight = new KruskalMinimumSpanningTree<>(g).getSpanningTree().getWeight();
+            double tourWeight = tour.getWeight();
+            assertTrue(2 * mstWeight >= tourWeight);
+        }
+    }
+
+    @Test
+    public void testStar()
+    {
+        SimpleWeightedGraph<String, DefaultWeightedEdge> g =
+            new SimpleWeightedGraph<>(DefaultWeightedEdge.class);
+        g.addVertex("1");
+        g.addVertex("2");
+        g.addVertex("3");
+        g.addVertex("4");
+        g.addVertex("5");
+        g.addVertex("6");
+
+        g.setEdgeWeight(g.addEdge("1", "2"), 1d);
+        g.setEdgeWeight(g.addEdge("1", "3"), 1d);
+        g.setEdgeWeight(g.addEdge("1", "4"), 1d);
+        g.setEdgeWeight(g.addEdge("1", "5"), 2d);
+        g.setEdgeWeight(g.addEdge("1", "6"), 2d);
+
+        g.setEdgeWeight(g.addEdge("2", "3"), 2d);
+        g.setEdgeWeight(g.addEdge("2", "4"), 1d);
+        g.setEdgeWeight(g.addEdge("2", "5"), 1d);
+        g.setEdgeWeight(g.addEdge("2", "6"), 2d);
+
+        g.setEdgeWeight(g.addEdge("3", "4"), 1d);
+        g.setEdgeWeight(g.addEdge("3", "5"), 2d);
+        g.setEdgeWeight(g.addEdge("3", "6"), 1d);
+
+        g.setEdgeWeight(g.addEdge("4", "5"), 1d);
+        g.setEdgeWeight(g.addEdge("4", "6"), 1d);
+
+        g.setEdgeWeight(g.addEdge("5", "6"), 1d);
+
+        GraphPath<String, DefaultWeightedEdge> tour =
+            new TwoApproxMetricTSP<String, DefaultWeightedEdge>().getTour(g);
+        assertHamiltonian(g, tour);
+
+        double mstWeight = new KruskalMinimumSpanningTree<>(g).getSpanningTree().getWeight();
+        double tourWeight = tour.getWeight();
+        assertTrue(2 * mstWeight >= tourWeight);
+
+    }
+
+    private static <V, E> void assertHamiltonian(UndirectedGraph<V, E> g, GraphPath<V, E> path)
+    {
+        // check that all vertices are visited
+        List<V> vertices = path.getVertexList();
+        Iterator<V> vIt = vertices.iterator();
+        V start = vIt.next();
+        Set<V> visited = new HashSet<>();
+        while (vIt.hasNext()) {
+            V v = vIt.next();
+            assertTrue(visited.add(v));
+            if (!vIt.hasNext()) {
+                assert (v.equals(start));
+            }
+        }
+        visited.add(start);
+        assertEquals(visited.size(), g.vertexSet().size());
+
+        // check that edges are valid
+        List<E> edges = path.getEdgeList();
+        if (edges.isEmpty()) {
+            return;
+        }
+
+        E prev = null;
+        E first = null, last = null;
+        Iterator<E> it = edges.iterator();
+        while (it.hasNext()) {
+            E cur = it.next();
+            if (prev == null) {
+                first = cur;
+            } else {
+                assertTrue(
+                    g.getEdgeSource(cur).equals(g.getEdgeSource(prev))
```

Comment:
This validation doesn't seem to be sufficient. Consider for instance the following edge sequence:
(3,4)(2,4)(5,4)
This is not a valid tour, but it would pass this test. A correct and shorter tour validation would be something like:
        List<V> list = new ArrayList<>();
        V v = path.getStartVertex();
        list.add(v);
        for (E e : edgeList) {
            v = Graphs.getOppositeVertex(g, e, v); //throws error if opposite vertex doesnt exist.
            list.add(v);
        }
       assertEquals(list.size(), g.vertexSet().size()); // check that all vertices are visited